components/fxa-client/src/telemetry.rs

```diff
@@ -2,12 +2,268 @@
+        // Default case: the ecosystem anon ID was generated during login.
+        if let Some(ecosystem_anon_id) = profile.ecosystem_anon_id {
+            return Ok(ecosystem_anon_id);
+        }
+        if !generate_placeholder {
+            return Err(ErrorKind::IllegalState("ecosystem_anon_id should be present").into());
+        }
+        // For older clients, we generate an ecosystem_user_id,
+        // persist it and then return ecosystem_anon_id.
+        let mut ecosystem_user_id = vec![0u8; 16];
+        rand::fill(&mut ecosystem_user_id)?;
+        // Will end up as a len 32 hex string.
+        let ecosystem_user_id = hex::encode(ecosystem_user_id);
+
+        let anon_id_key = self.fetch_random_ecosystem_anon_id_key()?;
+        let ecosystem_anon_id = jwcrypto::encrypt_to_jwe(
+            &ecosystem_user_id.as_bytes(),
+            EncryptionParameters::ECDH_ES {
+                enc: EncryptionAlgorithm::A256GCM,
+                peer_jwk: &anon_id_key,
+            },
+        )?;
+
+        let token = self.get_access_token(scopes::PROFILE_WRITE, None)?.token;
+        if let Err(err) =
+            self.client
+                .ecosystem_anon_id(&self.state.config, &token, &ecosystem_anon_id)
+        {
+            if let ErrorKind::RemoteError { code: 412, .. } = err.kind() {
+                // Another client beat us, fetch the new ecosystem_anon_id.
+                return self.get_ecosystem_anon_id_helper(false);
+            }
+        }
+
+        // Persist the unencrypted ecosystem_user_id for possible future use.
+        self.state.ecosystem_user_id = Some(ecosystem_user_id);
+        Ok(ecosystem_anon_id)
+    }
+
+    fn fetch_random_ecosystem_anon_id_key(&self) -> Result<Jwk> {
+        let config = self.client.fxa_client_configuration(&self.state.config)?;
+        let keys = config
+            .ecosystem_anon_id_keys
+            .ok_or_else(|| ErrorKind::NoAnonIdKey)?;
+        let mut rng = rand_rccrypto::RcCryptoRng;
+        Ok(keys
+            .choose(&mut rng)
+            .ok_or_else(|| ErrorKind::NoAnonIdKey)?
+            .clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{http_client::*, oauth::AccessTokenInfo, Config};
+    use jwcrypto::{ec::ECKeysParameters, JwkKeyParameters};
+    use std::sync::Arc;
+
+    fn fxa_setup() -> FirefoxAccount {
+        let config = Config::stable_dev("12345678", "https://foo.bar");
+        let mut fxa = FirefoxAccount::with_config(config);
+        fxa.add_cached_token(
+            "profile",
+            AccessTokenInfo {
+                scope: "profile".to_string(),
+                token: "profiletok".to_string(),
+                key: None,
+                expires_at: u64::max_value(),
+            },
+        );
+        fxa.add_cached_token(
+            "profile:write",
+            AccessTokenInfo {
+                scope: "profile".to_string(),
+                token: "profiletok".to_string(),
```

Comment:
testing nit: if you make this `"profilewritetok"` is similar, you might get some extra checks that the correct token is being used at the correct time.